rust_snuba/rust_arroyo/src/processing/strategies/reduce.rs

```diff
@@ -0,0 +1,195 @@
+use crate::processing::strategies::{CommitRequest, MessageRejected, ProcessingStrategy};
+use crate::types::{AnyMessage, InnerMessage, Message, Partition};
+use std::collections::BTreeMap;
+use std::time::{Duration, SystemTime};
+use std::mem;
+
+struct BatchState<T, TResult> {
+    value: Option<TResult>,
+    accumulator: fn(TResult, T) -> TResult,
+    offsets: BTreeMap<Partition, u64>,
+    batch_start_time: SystemTime,
+    message_count: usize,
+    is_complete: bool,
+}
+
+impl<T: Clone, TResult: Clone> BatchState<T, TResult> {
+    fn new(initial_value: TResult, accumulator: fn(TResult, T) -> TResult) -> BatchState<T, TResult> {
+        BatchState {
+            value: Some(initial_value),
+            accumulator,
+            offsets: Default::default(),
+            batch_start_time: SystemTime::now(),
+            message_count: 0,
+            is_complete: false,
+        }
+    }
+
+    fn add(&mut self, message: Message<T>) {
+        let mut tmp = None;
+        mem::swap(&mut tmp, &mut self.value);
+        self.value = Some((self.accumulator)(tmp.unwrap(), message.payload()));
+        self.message_count += 1;
+
+        for (partition, offset) in message.committable() {
+            self.offsets.insert(partition, offset);
+        }
+    }
+}
+
+pub struct Reduce<T, TResult> {
+    next_step: Box<dyn ProcessingStrategy<TResult>>,
+    accumulator: fn(TResult, T) -> TResult,
+    initial_value: TResult,
+    max_batch_size: usize,
+    max_batch_time: Duration,
+    batch_state: BatchState<T, TResult>,
```

Comment:
I would suggest to have `batch_state: None`, then lazily initializing `BatchState to `Some()` when a new batch is started. I believe this will allow you to use `self.batch_state.take()` when flushing (which avoids clones)

Reply:
The main reason I was cloning was for the `MessageRejected` case... where we would need to put back the batch state. Maybe we could send it back with `MessageRejected`?